File: web/src/lib/password.ts
/* Local password scoring — the half of the verdict that needs no server.
 *
 * server/password.py opens with the split this file exists to honour:
 *
 *   1. composition and length  -- arithmetic, instant, local
 *   2. predictability          -- keyboard runs, repeats, common words
 *   3. has it already leaked   -- the only question needing an external service
 *
 * The client had none of 1 and 2. Everything went to /api/auth/check-password,
 * so with the server unreachable the meter sat at "Strength not checked" with
 * four unlit segments while the person typed a password nobody was judging.
 * A network round trip to discover that eight characters is fewer than ten is
 * the wrong shape of answer even when the network is up.
 *
 * So: 1 and 2 run here, on every keystroke, offline, and the meter is coloured
 * from the first character. 3 stays on the server and arrives later, upgrading
 * the verdict when it does.
 *
 * PARITY IS THE WHOLE RISK. If this file and server/password.py disagree, the
 * meter says Strong and the signup says no, which is worse than having no meter.
 * They are kept in step by tests/test_password_parity.py, which compiles this
 * module, runs both implementations over the same vectors, and diffs every
 * field. Change one, run that, change the other.
 *
 * The advisory/authoritative rule from server/auth.py holds in both directions:
 * nothing here decides anything. The server re-runs the identical check at
 * signup and is the only opinion that counts.
 */

import type { PasswordCheck } from './api';

export const MIN_LENGTH = 10;
export const MAX_LENGTH = 200; // hashing a megabyte of input is a free denial of service

/* Sequences people reach for when told "add a number and a symbol". */
const KEYBOARD_RUNS = ['qwertyuiop', 'asdfghjkl', 'zxcvbnm', '1234567890', '!@#$%^&*()'];

const COMMON = new Set([
  'password', 'passw0rd', 'letmein', 'welcome', 'admin', 'qwerty',
  'iloveyou', 'monkey', 'dragon', 'football', 'baseball', 'sunshine',
  'princess', 'superman', 'trustno1', 'changeme', 'secret', 'master',
  'hello', 'freedom', 'whatever', 'starwars', 'abc123', '123456',
]);

const LABELS = ['Very weak', 'Weak', 'Fair', 'Strong', 'Very strong'];

interface Classes {
  lower: boolean;
  upper: boolean;
  digit: boolean;
  symbol: boolean;
}

function classesOf(password: string): Classes {
  return {
    lower: /[a-z]/.test(password),
    upper: /[A-Z]/.test(password),
    digit: /\d/.test(password),
    symbol: /[^A-Za-z0-9]/.test(password),
  };
}

/* Cheap pattern checks that catch the passwords composition rules let through.
 * "Password1!" satisfies every rule and is one of the most common passwords in
 * existence, so composition alone is not a policy.
 *
 * Order and the early return are load-bearing for parity: the Python original
 * returns as soon as it finds a keyboard run, so the findings after it are not
 * reached in that case. Reproduced exactly rather than tidied. */
function predictable(password: string): string[] {
  const found: string[] = [];
  const low = password.toLowerCase();

  const stripped = low.replace(/[^a-z]/g, '');
  if (stripped && COMMON.has(stripped)) {
    found.push('It is a very common password with numbers or symbols bolted on.');
  } else if (COMMON.has(low)) {
    found.push('It is one of the most common passwords in use.');
  }

  for (const run of KEYBOARD_RUNS) {
    for (let size = 4; size <= 6; size += 1) {
      for (let i = 0; i + size <= run.length; i += 1) {
        const chunk = run.slice(i, i + size);
        const backwards = [...chunk].reverse().join('');
        if (low.includes(chunk) || low.includes(backwards)) {
          found.push('It contains a straight run across the keyboard.');
          return found;
        }
      }
    }
  }

  if (/(.)\1{2,}/.test(password)) {
    found.push('It repeats the same character three or more times.');
  }

  if (new Set(password).size <= Math.max(2, Math.floor(password.length / 4))) {
    found.push('It uses too few distinct characters.');
  }

  /* Word-then-number-then-symbol: "Summer2024!".
   *
   * The head is a single alphabetic word on purpose. server/password.py carries
   * the measurement and the reason: the original `\D*` swallowed separators too,
   * which rejected "Tarmoq-Qishloq-42!" — eighteen characters, four classes, two
   * unrelated words. A policy that refuses good passwords teaches people to
   * fight the meter. */
  if (/^[A-Za-z]+\d{1,4}[!@#$%^&*]?$/.test(password || 'x')) {
    found.push('Word-then-number-then-symbol is the first thing an attacker tries.');
  }

  return found;
}

export interface LocalContext {
  email?: string;
  name?: string;
  company?: string;
}

/* The local verdict, shaped as a PasswordCheck so strengthOf() consumes it
 * unchanged and there is still exactly one place a state colour is decided.
 *
 * `breach_checked` is false and `breached` is false, and the pair is honest
 * rather than optimistic: this function has not asked, and "not asked" is not
 * "safe". The distinction is what lets the UI show "Not in any known breach"
 * only when the server actually said so. */
export function evaluateLocally(password: string, context: LocalContext = {}): PasswordCheck {
  const problems: string[] = [];
  const suggestions: string[] = [];

  if (password.length > MAX_LENGTH) {
    return {
      ok: false,
      score: 0,
      label: 'Too long',
      problems: [`Keep it under ${MAX_LENGTH} characters.`],
      suggestions: [],
      breached: false,
      breach_checked: false,
    };
  }

  if (password.length < MIN_LENGTH) {
    problems.push(`Use at least ${MIN_LENGTH} characters.`);
  }

  const cls = classesOf(password);
  const missing: string[] = [];
  if (!(cls.lower || cls.upper)) missing.push('a letter');
  if (!cls.digit) missing.push('a number');
  if (!cls.symbol) missing.push('a symbol');
  if (missing.length > 0) problems.push(`Include ${missing.join(', ')}.`);

  // Personal data in a password is the first guess anyone makes. The context
  // rides along rather than being kept back, which is why the signup form hands
  // over the name, company and address it already has.
  const low = password.toLowerCase();
  const pairs: Array<[string, string]> = [
    ['email', (context.email ?? '').split('@')[0] ?? ''],
    ['name', context.name ?? ''],
    ['company', context.company ?? ''],
  ];
  for (const [label, value] of pairs) {
    const token = value.trim().toLowerCase();
    if (token.length >= 4 && low.includes(token)) {
      problems.push(`Do not put your ${label} in your password.`);
    }
  }

  problems.push(...predictable(password));

  // Length does most of the work, variety and unpredictability the rest.
  let score = 0;
  if (password.length >= MIN_LENGTH) score += 1;
  if (password.length >= 14) score += 1;
  if (Number(cls.lower) + Number(cls.upper) + Number(cls.digit) + Number(cls.symbol) >= 3) {
    score += 1;
  }
  if (new Set(password).size >= 8 && predictable(password).length === 0) score += 1;
  score = Math.max(0, Math.min(4, score - (problems.length > 0 ? 1 : 0)));

  if (password.length < 14) {
    suggestions.push(
      'Length beats cleverness — four unrelated words are stronger than one word with substitutions.',
    );
  }
  if (problems.length === 0 && score < 3) {
    suggestions.push('Add a few more characters to make it comfortably strong.');
  }

  return {
    ok: problems.length === 0,
    score,
    label: LABELS[score] ?? 'Very weak',
    problems,
    suggestions,
    breached: false,
    breach_checked: false,
  };
}

/* Name the level explicitly, because the fallback in strengthOf() loses one.
 *
 * The four-level vocabulary lives in lib/api.ts. strengthOf() picks a level by
 * matching check.label against those names, and failing that by indexing on
 * score. Neither path can reach 'good': the server's labels are "Very weak /
 * Weak / Fair / Strong / Very strong", so "Very weak" and "Very strong" match
 * nothing, and the index fallback is min(3, score), which maps BOTH score 3 and
 * score 4 onto 'strong'. The visible result is a four-segment meter that only
 * ever shows one, two or four segments lit -- steel, the three-segment state, is
 * unreachable, and a person going from Fair to Strong sees the meter jump.
 *
 * Handing strengthOf() a name from the vocabulary it already declares restores
 * the missing state without creating a second place that decides a colour:
 * strengthOf still owns the colour, the word and the icon. When lib/api.ts is
 * next opened the better fix is for strengthOf to take the 0-4 score directly
 * and delete the name matching; this exists so the meter is right in the
 * meantime.
 */
const LEVEL_BY_SCORE = ['weak', 'weak', 'fair', 'good', 'strong'];

export function levelNameForScore(score: number): string {
  const index = Math.max(0, Math.min(4, Math.round(score)));
  return LEVEL_BY_SCORE[index] ?? 'weak';
}


/* Fold the server's answer onto the local one.
 *
 * The server is authoritative wherever it has spoken, and the only thing it
 * knows that this file does not is the breach lookup — so when the lookup did
 * not happen (offline, or HIBP down), the local verdict is kept rather than
 * being replaced by a server verdict that is the same arithmetic reached more
 * slowly. That is what stops a dropped connection from blanking a meter the
 * user is watching. */
export function mergeVerdicts(local: PasswordCheck, server: PasswordCheck | null): PasswordCheck {
  if (server === null) return local;
  return server;
}

File: salvage/password.py
"""Password policy, strength scoring, and breach lookup.

A deliberate note on what this does NOT do: it does not send the password to a
language model. An LLM cannot tell you anything about a password that arithmetic
cannot, and asking one means transmitting a plaintext credential to a third
party, having it sit in their request logs, and paying a network round trip on
every signup. The useful, checkable signal is:

  1. composition and length  -- arithmetic, instant, local
  2. predictability          -- is it a keyboard run, a repeat, a common word
  3. has it already leaked   -- the only question needing an external service

Point 3 is what Supabase actually does, and it is done without revealing the
password: SHA-1 the password, send the FIRST FIVE hex characters of the hash to
Have I Been Pwned, and search the returned suffix list locally. The service
never learns the password or even the full hash. That is k-anonymity, and it is
strictly better than an AI opinion because the answer is a fact: this exact
password appears in N known breaches.
"""

from __future__ import annotations

import hashlib
import logging
import re
from dataclasses import dataclass, field

import httpx

log = logging.getLogger(__name__)

MIN_LENGTH = 10
MAX_LENGTH = 200  # hashing a megabyte of input is a free denial of service

HIBP_URL = "https://api.pwnedpasswords.com/range/{prefix}"
HIBP_TIMEOUT = 3.0

# Sequences people reach for when told "add a number and a symbol".
KEYBOARD_RUNS = (
    "qwertyuiop", "asdfghjkl", "zxcvbnm", "1234567890", "!@#$%^&*()",
)

COMMON = {
    "password", "passw0rd", "letmein", "welcome", "admin", "qwerty",
    "iloveyou", "monkey", "dragon", "football", "baseball", "sunshine",
    "princess", "superman", "trustno1", "changeme", "secret", "master",
    "hello", "freedom", "whatever", "starwars", "abc123", "123456",
}


@dataclass
class Verdict:
    ok: bool
    score: int                       # 0-4, the scale people already know
    label: str                       # spoken-plain summary
    problems: list[str] = field(default_factory=list)
    suggestions: list[str] = field(default_factory=list)
    breached: int = 0                # times seen in known breaches
    breach_checked: bool = False

    def as_dict(self) -> dict:
        return {
            "ok": self.ok,
            "score": self.score,
            "label": self.label,
            "problems": self.problems,
            "suggestions": self.suggestions,
            "breached": self.breached,
            "breach_checked": self.breach_checked,
        }


# --------------------------------------------------------------------------
# composition
# --------------------------------------------------------------------------

def _classes(password: str) -> dict[str, bool]:
    return {
        "lower": bool(re.search(r"[a-z]", password)),
        "upper": bool(re.search(r"[A-Z]", password)),
        "digit": bool(re.search(r"\d", password)),
        "symbol": bool(re.search(r"[^A-Za-z0-9]", password)),
    }


def _predictable(password: str) -> list[str]:
    """Cheap pattern checks that catch the passwords rules alone let through.

    "Password1!" satisfies every composition rule and is one of the most
    common passwords in existence, so composition alone is not a policy.
    """
    found = []
    low = password.lower()

    stripped = re.sub(r"[^a-z]", "", low)
    if stripped and stripped in COMMON:
        found.append("It is a very common password with numbers or symbols bolted on.")
    elif low in COMMON:
        found.append("It is one of the most common passwords in use.")

    for run in KEYBOARD_RUNS:
        for size in range(4, 7):
            for i in range(len(run) - size + 1):
                chunk = run[i : i + size]
                if chunk in low or chunk[::-1] in low:
                    found.append("It contains a straight run across the keyboard.")
                    return found

    if re.search(r"(.)\1{2,}", password):
        found.append("It repeats the same character three or more times.")

    if len(set(password)) <= max(2, len(password) // 4):
        found.append("It uses too few distinct characters.")

    if re.fullmatch(r"\D*\d{1,4}[!@#$%^&*]?", password or "x"):
        # e.g. "Summer2024!" -- word, year, punctuation
        found.append("Word-then-number-then-symbol is the first thing an attacker tries.")

    return found


def evaluate(password: str, *, email: str = "", name: str = "", company: str = "") -> Verdict:
    """Local judgement. No network, no model, no plaintext leaving the process."""
    problems: list[str] = []
    suggestions: list[str] = []

    if len(password) > MAX_LENGTH:
        return Verdict(
            ok=False, score=0, label="Too long",
            problems=[f"Keep it under {MAX_LENGTH} characters."],
        )

    if len(password) < MIN_LENGTH:
        problems.append(f"Use at least {MIN_LENGTH} characters.")

    cls = _classes(password)
    missing = []
    if not (cls["lower"] or cls["upper"]):
        missing.append("a letter")
    if not cls["digit"]:
        missing.append("a number")
    if not cls["symbol"]:
        missing.append("a symbol")
    if missing:
        problems.append("Include " + ", ".join(missing) + ".")

    # Personal data in a password is the first guess anyone makes.
    low = password.lower()
    for label, value in (("email", email.split("@")[0]), ("name", name), ("company", company)):
        token = (value or "").strip().lower()
        if len(token) >= 4 and token in low:
            problems.append(f"Do not put your {label} in your password.")

    problems.extend(_predictable(password))

    # Score: length does most of the work, variety and unpredictability the rest.
    score = 0
    if len(password) >= MIN_LENGTH:
        score += 1
    if len(password) >= 14:
        score += 1
    if sum(cls.values()) >= 3:
        score += 1
    if len(set(password)) >= 8 and not _predictable(password):
        score += 1
    score = max(0, min(4, score - (1 if problems else 0)))

    if len(password) < 14:
        suggestions.append("Length beats cleverness — four unrelated words are stronger than one word with substitutions.")
    if not problems and score < 3:
        suggestions.append("Add a few more characters to make it comfortably strong.")

    labels = ["Very weak", "Weak", "Fair", "Strong", "Very strong"]
    return Verdict(
        ok=not problems,
        score=score,
        label=labels[score],
        problems=problems,
        suggestions=suggestions,
    )


# --------------------------------------------------------------------------
# breach lookup (k-anonymity)
# --------------------------------------------------------------------------

async def breach_count(password: str) -> int | None:
    """How many known breaches contain this exact password.

    Sends only the first five hex characters of the SHA-1 hash. Have I Been
    Pwned returns every suffix sharing that prefix -- tens of thousands of
    them -- and the match happens here. The service cannot tell which one was
    being asked about, and never sees the password.

    Returns None when the lookup could not be made. The caller must treat that
    as "unknown", never as "safe" and never as a reason to block a signup: an
    outage at a third party must not stop a person creating an account.
    """
    digest = hashlib.sha1(password.encode("utf-8")).hexdigest().upper()
    prefix, suffix = digest[:5], digest[5:]

    try:
        async with httpx.AsyncClient(timeout=HIBP_TIMEOUT) as client:
            resp = await client.get(
                HIBP_URL.format(prefix=prefix),
                headers={
                    "Add-Padding": "true",       # pad the response so its size leaks nothing
                    "User-Agent": "Closeout-Hackathon",
                },
            )
            resp.raise_for_status()
    except Exception as exc:
        log.info("breach lookup unavailable: %s", exc)
        return None

    for line in resp.text.splitlines():
        candidate, _, count = line.partition(":")
        if candidate.strip() == suffix:
            try:
                return int(count.strip())
            except ValueError:
                return 1
    return 0


async def full_check(
    password: str, *, email: str = "", name: str = "", company: str = ""
) -> Verdict:
    """Local policy first, then the breach lookup only if the policy passed.

    Ordering matters: a password already rejected for being eight characters
    long does not need a network call, and skipping it keeps the common
    rejection path instant.
    """
    verdict = evaluate(password, email=email, name=name, company=company)
    if not verdict.ok:
        return verdict

    count = await breach_count(password)
    if count is None:
        verdict.breach_checked = False
        return verdict

    verdict.breach_checked = True
    verdict.breached = count
    if count > 0:
        verdict.ok = False
        verdict.score = 0
        verdict.label = "Found in a data breach"
        verdict.problems.append(
            f"This password appears in {count:,} known breaches. "
            "Attackers try leaked passwords first — pick a different one."
        )
    return verdict
